source/matrix_free/vector_data_exchange.cc

```diff
@@ -316,6 +318,1057 @@ namespace internal
+          }
+
+        // send data to remote processes
+        for (unsigned int i = 0, k = 0; i < import_targets_data.size(); i++)
+          {
+            for (unsigned int j = import_indices_data.first[i];
+                 j < import_indices_data.first[i + 1];
+                 j++)
+              for (unsigned int l = 0; l < import_indices_data.second[j].second;
+                   l++, k++)
+                temporary_storage[k] =
+                  data_this[import_indices_data.second[j].first + l];
+
+            // send data away
+            MPI_Isend(temporary_storage.data() +
+                        import_targets_data[i].second.first,
+                      import_targets_data[i].second.second,
+                      Utilities::MPI::internal::mpi_type_id(data_this.data()),
+                      import_targets_data[i].first,
+                      communication_channel + 3,
+                      comm,
+                      requests.data() + sm_import_ranks.size() +
+                        sm_ghost_ranks.size() + ghost_targets_data.size() + i);
+          }
+#endif
+      }
+
+
+
+      template <typename Number>
+      void
+      Full::export_to_ghosted_array_finish_impl(
+        const ArrayView<const Number> &             data_this,
+        const std::vector<ArrayView<const Number>> &data_others,
+        const ArrayView<Number> &                   ghost_array,
+        std::vector<MPI_Request> &                  requests) const
+      {
+        (void)data_this;
+
+#ifndef DEAL_II_WITH_MPI
+        Assert(false, ExcNeedsMPI());
+
+        (void)data_others;
+        (void)ghost_array;
+        (void)requests;
+#else
+
+        AssertDimension(requests.size(),
+                        sm_import_ranks.size() + sm_ghost_ranks.size() +
+                          ghost_targets_data.size() +
+                          import_targets_data.size());
+
+        const auto split =
+          [&](const unsigned int i) -> std::pair<unsigned int, unsigned int> {
+          AssertIndexRange(i,
+                           (sm_ghost_ranks.size() + ghost_targets_data.size()));
+
+          if (i < sm_ghost_ranks.size())
+            return {0, i};
+          else
+            return {1, i - sm_ghost_ranks.size()};
+        };
+
+        for (unsigned int c = 0;
+             c < sm_ghost_ranks.size() + ghost_targets_data.size();
+             c++)
+          {
+            int i;
+            MPI_Waitany(sm_ghost_ranks.size() + ghost_targets_data.size(),
+                        requests.data() + sm_import_ranks.size(),
+                        &i,
+                        MPI_STATUS_IGNORE);
+
+            const auto s = split(i);
+            i            = s.second;
+
+            if (s.first == 0)
+              {
+                const Number *__restrict__ data_others_ptr =
+                  data_others[sm_ghost_ranks[i]].data();
+                Number *__restrict__ data_this_ptr = ghost_array.data();
```

Comment:
Can you use 
```suggestion
                const Number *DEAL_II_RESTRICT data_others_ptr =
                  data_others[sm_ghost_ranks[i]].data();
                Number *DEAL_II_RESTRICT data_this_ptr = ghost_array.data();
```
to make sure this works with all compilers?